developer/frameworks/articles/dependency-injection/02-cdi-dependency-injection.markdown

```diff
@@ -0,0 +1,142 @@
+---
+header-id: cdi-dependency-injection
+---
+
+# CDI Dependency Injection
+
+[TOC levels=1-4]
+
+Portlet 3.0 (see [JSR 362](https://jcp.org/en/jsr/detail?id=362)) supports using
+Contexts and Dependency Injection (CDI). It enables you to create and use
+injectable classes (CDI beans) in your portlet. It also provides injectable
+portlet artifacts called [Portlet Predefined
+Beans](/docs/7-2/reference/-/knowledge_base/r/cdi-portlet-predefined-beans).
+They give a portlet's CDI beans access to the portlet configuration,
+preferences, requests, responses, and more. Here you'll create and use CDI beans
+and use Portlet Predefined Beans in your portlet. 
+
+1.  Create a portlet WAR project, if you haven't created one 
+    already. 
+
+    -   [Bean Portlet](/docs/7-2/appdev/-/knowledge_base/a/bean-portlet) 
+    -   [JSF Portlet](/docs/7-2/appdev/-/knowledge_base/a/jsf-portlet) 
```

Comment:
@jhinkey @sez11a At this time, we can't support JSF portlets that leverage CDI (and have a src/main/webapp/WEB-INF/beans.xml) as "bean portlets" -- in other words, they need to go through the traditional portlet registration and not be regarded as bean portlets. It also means they have to include WEB-INF/lib/weld-servlet.jar and can't rely on weld-osgi-bundle.jar that ships with Liferay as part of the CDI+OSGi integration feature. We hope to remedy this in the future.

/cc @stiemannkj1